File: 6_semester/Интернет вещи/ЛР2/things.py
"""
Цифровые двойники «Умный дом» (вариант 1).
Иерархия согласована с первой лабораторной работой.
"""
from __future__ import annotations

import abc
import logging

logger = logging.getLogger(__name__)


class Thing(abc.ABC):
    """Базовая абстрактная «вещь» IoT."""

    def __init__(self, id_: str, name: str, is_online: bool = True) -> None:
        self.id = id_
        self.name = name
        self.is_online = is_online
        logger.info("Создан цифровой двойник: %s (%s)", name, id_)

    @abc.abstractmethod
    def get_status(self) -> str:
        """Текстовое состояние для мониторинга."""

    def ping(self) -> None:
        """Проверка связи; достаточно сообщения в лог (требования ЛР2)."""
        logger.info("Метод ping вызван у вещи %s", self.name)


class TemperatureSensor(Thing):
    """Датчик температуры."""

    def __init__(self, id_: str, name: str, initial_c: float = 22.0) -> None:
        super().__init__(id_, name)
        self.current_temperature_c = initial_c

    def set_temperature(self, value: float) -> None:
        logger.info("TemperatureSensor.set_temperature запущен, вещь=%s", self.name)
        self.current_temperature_c = value

    def get_temperature(self) -> float:
        logger.info("TemperatureSensor.get_temperature запущен, вещь=%s", self.name)
        return self.current_temperature_c

    def get_status(self) -> str:
        return f"{self.name}: {self.current_temperature_c:.1f} °C"


class LightController(Thing):
    """Управление освещением."""

    def __init__(self, id_: str, name: str) -> None:
        super().__init__(id_, name)
        self.brightness_percent = 0
        self.is_enabled = False

    def set_light(self, enabled: bool, brightness: int) -> None:
        logger.info("LightController.set_light запущен, вещь=%s", self.name)
        self.is_enabled = enabled
        self.brightness_percent = max(0, min(100, brightness))

    def get_status(self) -> str:
        state = "ВКЛ" if self.is_enabled else "ВЫКЛ"
        return f"{self.name}: {state}, яркость={self.brightness_percent}%"


class SmartLock(Thing):
    """Умный замок."""

    def __init__(self, id_: str, name: str) -> None:
        super().__init__(id_, name)
        self.locked = True

    def set_locked(self, locked: bool) -> None:
        logger.info("SmartLock.set_locked запущен, вещь=%s", self.name)
        self.locked = locked

    def get_status(self) -> str:
        state = "ЗАКРЫТ" if self.locked else "ОТКРЫТ"
        return f"{self.name}: {state}"


class DeviceDataStore:
    """Упрощённое хранилище записей мониторинга (сущность данных)."""

    def __init__(self) -> None:
        self._by_id: dict[str, list[str]] = {}

    def save_record(self, thing: Thing, payload: str) -> None:
        logger.info("DeviceDataStore.save_record запущен для %s", thing.id)
        self._by_id.setdefault(thing.id, []).append(payload)


class MainControlUnit:
    """Центральный логический узел."""

    def __init__(self, store: DeviceDataStore) -> None:
        self._store = store
        self._things: list[Thing] = []

    def register_thing(self, thing: Thing) -> None:
        logger.info("MainControlUnit.register_thing запущен: %s", thing.name)
        self._things.append(thing)

    def collect_monitoring_snapshot(self) -> None:
        logger.info("MainControlUnit.collect_monitoring_snapshot запущен")
        for t in self._things:
            line = f"{t.get_status()}"
            self._store.save_record(t, line)

File: 6_semester/Интернет вещи/Lab4/src/things.ts
export type CommandResult = { ok: true } | { ok: false; error: string };

export abstract class Thing {
    constructor(
        public id: string,
        public name: string,
        public isOnline: boolean = true
    ) {}

    protected abstract emulate(): void
    protected abstract buildPayload(): Record<string, unknown>;
    
    connect(): Record<string, unknown> {
        console.log(`[Thing.connect] ${this.constructor.name} (${this.id})`);
        this.emulate();
        return this.buildPayload();
    }

    /** Снимок состояния без эмуляции (после команды). */
    snapshot(): Record<string, unknown> {
        return this.buildPayload();
    }

    abstract getStatus(): string;
}

export interface IParent {
    render(statuses: string[]): string;
}

export interface IChild {
    render(statuses: string[]): string;
}

export class MainControlUnit {
    constructor(
        public readonly things: Thing[],
        public readonly dataStore: DeviceDataStore
    ) {}

    registerThing(thing: Thing): void {
        console.log(`[MainControlUnit.registerThing] ${thing.id}`);
        this.things.push(thing);
    }
    collectMonitoringData(): void {
        console.log(`[MainControlUnit.collectMonitoringData]` );
        for (const thing of this.things) {
            const line = `[${new Date().toISOString()}] ${thing.getStatus()}`
            this.dataStore.saveRecord(thing, line)
        }
    }
    getAllStatuses(): string[] {
        console.log(`[MainControlUnit.getAllStatuses]`);
        return this.things.map((t) => t.getStatus());
    }
}

export class DeviceDataStore {
    private readonly dataByThingId = new Map<string, string[]>();

    saveRecord(thing: Thing, payload: string): void {
        console.log(`[DeviceDataStore.saveRecord] ${thing.id}` );
        const list = this.dataByThingId.get(thing.id) ?? [];
        list.push(payload);
        this.dataByThingId.set(thing.id, list);
    }

    getRecords(thingId: string): string[] {
        console.log(`[DeviceDataStore.getRecords] ${thingId}` );
        return this.dataByThingId.get(thingId) ?? [];
    }
    
}

export class TemperatureSensor extends Thing {
    private currentTemperatureC: number;

    constructor(id: string, name: string, initialTemperatureC: number) {
        super(id, name, true);
        this.currentTemperatureC = initialTemperatureC;
    }

    setTemperature(valueC: number): void {
        console.log(`[TemperatureSensor.setTemperature] ${this.id}`);
        this.currentTemperatureC = valueC;
    }

    getTemperature(): number {
        return this.currentTemperatureC;
    }

    /** ЛР4: параметры из query (аналог request.args). */
    applyCommand(query: Record<string, string | undefined>): CommandResult {
        const raw = query.value ?? query.temperature;
        if (raw === undefined || raw === "") {
            return { ok: false, error: "Ожидается параметр value (или temperature)" };
        }
        const n = Number(raw);
        if (Number.isNaN(n)) {
            return { ok: false, error: "value должно быть числом" };
        }
        this.setTemperature(n);
        console.log(`[TemperatureSensor.applyCommand] ${this.id} -> ${n}`);
        return { ok: true };
    }

    protected emulate(): void {
        const delta = (Math.random() - 0.5) * 1.2;
        this.currentTemperatureC = Math.round((this.currentTemperatureC + delta) * 10) / 10;
    }

    protected buildPayload(): Record<string, unknown> {
        return {
          id: this.id,
          name: this.name,
          value: this.currentTemperatureC,
          unit: "°C"
        };
    }
    

    getStatus(): string {
        console.log(`[TemperatureSensor.getStatus] ${this.id}`);
        return `${this.name}: ${this.currentTemperatureC.toFixed(1)} °C`
    }
}

export class LightController extends Thing {
    private brightnessPercent: number = 0;
    private isEnabled: boolean = false;

    setLight(isEnabled: boolean, brightnessPercent: number): void {
        console.log(`[LightController.setLight] ${this.id}`);
        this.isEnabled = isEnabled;
        this.brightnessPercent = Math.max(0, Math.min(100, brightnessPercent))
    }

    applyCommand(query: Record<string, string | undefined>): CommandResult {
        const enabledRaw = query.enabled ?? query.on;
        const brightRaw = query.brightness ?? query.brightnessPercent;
        if (enabledRaw === undefined || brightRaw === undefined || brightRaw === "") {
            return { ok: false, error: "Ожидаются enabled (true/false) и brightness (0–100)" };
        }
        const enabled = enabledRaw === "true" || enabledRaw === "1";
        const brightness = Number(brightRaw);
        if (Number.isNaN(brightness)) {
            return { ok: false, error: "brightness должно быть числом" };
        }
        this.setLight(enabled, brightness);
        console.log(`[LightController.applyCommand] ${this.id} enabled=${enabled} brightness=${brightness}`);
        return { ok: true };
    }

    protected emulate(): void {
        const delta = Math.floor((Math.random() - 0.5) * 12);
        this.brightnessPercent = Math.max(0, Math.min(100, this.brightnessPercent + delta));
    }

    protected buildPayload(): Record<string, unknown> {
        return {
          id: this.id,
          name: this.name,
          enabled: this.isEnabled,
          brightnessPercent: this.brightnessPercent
        };
    }

    getStatus(): string {
        console.log(`[LightController.getStatus] ${this.id}`);
        return `${this.name}: ${this.isEnabled ? "ВКЛ" : "ВЫКЛ"}, яркость=${this.brightnessPercent}%`;
    }
}

export class SmartLock extends Thing{
    private isLocked: boolean = true;

    setLocked(value: boolean): void {
        console.log(`[SmartLock.setLocked] ${this.id}`);
         this.isLocked = value;
    }

    applyCommand(query: Record<string, string | undefined>): CommandResult {
        const raw = query.locked ?? query.lock;
        if (raw === undefined || raw === "") {
            return { ok: false, error: "Ожидается locked (true/false)" };
        }
        if (raw !== "true" && raw !== "false" && raw !== "1" && raw !== "0") {
            return { ok: false, error: "locked должно быть true или false" };
        }
        const locked = raw === "true" || raw === "1";
        this.setLocked(locked);
        console.log(`[SmartLock.applyCommand] ${this.id} locked=${locked}`);
        return { ok: true };
    }

    protected emulate(): void {
        if (Math.random() < 0.08) this.isLocked = !this.isLocked;
      }
    
    protected buildPayload(): Record<string, unknown> {
        return {
          id: this.id,
          name: this.name,
          locked: this.isLocked
        };
    }

    getStatus(): string {
        console.log(`[SmartLock.getStatus] ${this.id}`);
        return `${this.name}: ${this.isLocked ? "ЗАКРЫТ" : "ОТКРЫТ"}`;
    }
}

export class ParentInterface implements IParent {
    render(statuses: string[]): string {
        console.log("[ParentInterface.render]");
        return ["Интерфейс родителя (полный доступ)", ...statuses].join("\n");
    }
}

export class ChildInterface implements IChild {
    render(statuses: string[]): string {
        console.log("[ChildInterface.render]");
        return ["Интерфейс ребенка (ограничен)", ...statuses].join("\n");
    }
}

export function runDemo(): {
    parentText: string;
    childText: string;
    sensorRecords: string[];
  } {
    const store = new DeviceDataStore();
    const things: Thing[] = [];
    const mcu = new MainControlUnit(things, store);
    const temp = new TemperatureSensor("sensor-1", "Температурный датчик", 22.5);
    const light = new LightController("light-1", "Свет в гостиной");
    const lock = new SmartLock("lock-1", "Замок входной двери");
    light.setLight(true, 65);
    lock.setLocked(true);
    mcu.registerThing(temp);
    mcu.registerThing(light);
    mcu.registerThing(lock);
    mcu.collectMonitoringData();
    const statuses = mcu.getAllStatuses();
    const parentText = new ParentInterface().render(statuses);
    const childText = new ChildInterface().render(statuses);
    return {
      parentText,
      childText,
      sensorRecords: store.getRecords("sensor-1")
    };
  }